CIP-0140/README.md

```diff
@@ -0,0 +1,1572 @@
+
+An [online simulator for Peras](https://peras-simulation.cardano-scaling.org/) is available.
+
+
+### Normative Peras specification in Agda
+
+The following formal, relational specification for Peras type utilizes [Agda 2.6.4.3](https://github.com/agda/agda/tree/v2.6.4.3). See [the Appendix](#typechecking-this-specification) for instruction on type-checking this specification with the Agda compiler and see [github:input-output-hk/peras-design](https://github.com/input-output-hk/peras-design/) for proofs and other modules related to this specification.
+
+```agda
+module README where
+```
+
+Most of the imports come from the [Agda Standard Library 2.0](https://github.com/agda/agda-stdlib/tree/v2.0).
+
+```agda
+open import Data.Bool using (Bool; if_then_else_; not; _∧_)
+open import Data.Empty using (⊥)
+open import Data.Fin using (Fin; pred)
+                     renaming (zero to fzero; suc to fsuc)
+open import Data.List using (List; any; concat; dropWhile; filter; head; map; mapMaybe; sum; []; _∷_; _++_)
+                      renaming (length to ∣_∣)
+open import Data.List.Membership.Propositional using (_∈_)
+open import Data.List.Relation.Unary.All using (All)
+open import Data.List.Relation.Unary.Any using (Any; any?; _─_)
+                                         renaming (_∷=_ to _∷ˡ=_)
+open import Data.Maybe using (Maybe; just; nothing)
+open import Data.Nat using (NonZero; Ordering; suc; ℕ; _≤_; _≥_; _>_; _≤?_; _<ᵇ_; _≤ᵇ_; _+_; _∸_; _*_; _/_; _%_)
+open import Data.Nat.Properties using (_≟_)
+open import Data.Product using (proj₁; proj₂; _×_; _,_)
+open import Data.Sum using (_⊎_)
+open import Data.Unit using (⊤)
+open import Function.Base using (_∘_)
+open import Relation.Binary using (DecidableEquality)
+open import Relation.Binary.PropositionalEquality using (_≡_; _≢_; cong)
+open import Relation.Nullary using (Dec; no; yes; ¬_; ⌊_⌋)
+open import Relation.Nullary.Decidable using (_×-dec_; ¬?)
+```
+
+Several additional imports come from the [IOG Agda Prelude v0.1.0.0](https://github.com/input-output-hk/iog-agda-prelude/releases/tag/v0.1.0.0).
+
+```agda
+open import Prelude.AssocList using (AssocList; set; _⁉_)
+open import Prelude.DecEq using (DecEq; _==_)
+open import Prelude.Default using (Default)
+open import Prelude.InferenceRules
+```
+
+#### Protocol parameters
+
+The seven *protocol parameters* are natural numbers.
+
+```agda
+record Params : Set where
+  field
+```
+
+The $U$ parameter is the duration of each voting round, measured in slots.
+
+```agda
+        U : ℕ
+```
+
+The $L$ parameter is the minimum age of a candidate block for being voted upon, measured in slots.
+
+```agda
+        L : ℕ
+```
+
+<span id="#expiration"/>The $A$ parameter is the maximum age for a certificate to be included in a block, measured in slots.
+
+```agda
+        A : ℕ
```

Comment:
$A$ is used in the condition $r - \mathsf{round}(\mathsf{cert}^\prime) \leq A$, where the right hand side is a number of rounds, but $A$ is measured in slots as stated here. This is incorrect (confirmed with @corettis), $A$ should rather be measured in rounds (and therefore be equal to $R$).

I guess the Agda specification could benefit from using some unit system to rule out these kinds of bugs ("natural number blindness").